datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java

```diff
@@ -0,0 +1,160 @@
+            reattachStack(reattached, detachedName);
+
+            if (clusterToReattach.hasExternalDatabase()) {
+                reattachExternalDatabase(reattached, detachedCrn);
+            }
+
+            return reattached;
+        });
+    }
+
+    /**
+     * Reattaches the internal SDX cluster by reverting its name and CRN to original values.
+     *
+     * Throws RuntimeException if cluster is not detached or could not revert name and CRN.
+     */
+    public SdxCluster reattachCluster(SdxCluster cluster) {
+        LOGGER.info("Started reattaching SDX cluster with ID: {}", cluster.getId());
+
+        if (!cluster.isDetached()) {
+            throw new RuntimeException("Attempting to reattach a cluster which was not detached!");
+        }
+
+        modifyClusterNameAndCrnForReattach(cluster);
+        cluster.setDetached(false);
+        cluster.setOriginalCrn(null);
+        SdxCluster saved = sdxClusterRepository.save(cluster);
+
+        LOGGER.info("Finished reattaching SDX cluster with ID: {}. Cluster now has name {} and crn {}.",
+                saved.getId(), saved.getName(), saved.getCrn());
+        return saved;
+    }
+
+    /**
+     * Gives the cluster being reattached its original name and CRN.
+     *
+     * Throws RuntimeException if the cluster's name is malformed from what is expected for a detached cluster.
+     */
+    private void modifyClusterNameAndCrnForReattach(SdxCluster cluster) {
+        String detachedCrn = cluster.getOriginalCrn();
+        cluster.setCrn(detachedCrn);
+        cluster.setClusterName(sdxDetachNameGenerator.generateOriginalNameFromDetached(cluster.getClusterName()));
+    }
+
+    /**
+     * Reattaches the backing stack of the SDX cluster by reverting its name and CRN.
+     *
+     * Can throw the following exceptions:
+     *      - NotFoundException : Internal stack is not found by the cluster's name.
+     *      - TransactionExecutionException : If updating the stack fails.
+     */
+    public void reattachStack(SdxCluster cluster, String originalName) {
+        LOGGER.info("Started reattaching stack of SDX cluster with ID: {}", cluster.getId());
+
+        String initiatorUserCrn = ThreadBasedUserCrnProvider.getUserCrn();
+        ThreadBasedUserCrnProvider.doAsInternalActor(() ->
+                stackV4Endpoint.updateNameAndCrn(
+                        0L, originalName, initiatorUserCrn, cluster.getClusterName(), cluster.getCrn()
+                )
+        );
+
+        LOGGER.info("Finished reattaching stack of SDX cluster with ID: {}. Now has name {} and crn {}.",
+                cluster.getId(), cluster.getName(), cluster.getCrn());
+    }
+
+    /**
+     * Reattaches the external database of the SDX cluster by reverting its CRN.
+     *
+     * Throws a NotFoundException if the database is not found for the cluster CRN and environment CRN.
+     */
+    public void reattachExternalDatabase(SdxCluster cluster, String detachedCrn) {
```

Comment:
detachStack in other class and reattachStack also this method and detachExternalDatabase has the same logic with different parameters, it is worth to generalize them.
in general SdxAttachService and SDxDeatchService have a lot in common

Reply:
Added a new component for this, let me know if you think it looks better now.

Reply:
yes, thank you!